### DISCUSSION TITLE
멀티 프로세스 대신 멀티 스레드를 사용하는 이유를 설명해주세요.

### DISCUSSION THREAD

**Discussion by eatingbug** [Team-Psyduck/tech-sharing#63] [General]
- 메모리 관리 측면에서 멀티 스레드가 멀티 프로세스보다 어떤 이점을 가지고 있는가요?

**Comment by kkg5** [comment]
멀티 스레드 방식은 컨텍스트 스위칭 비용이 적고 자원을 공유하기 쉽습니다.

그러나 자원을 공유하고 있을 때 발생할 수 있는 동시성 문제와 같은 것들에 유의해야 합니다.

**Reply by eatingbug** [reply]
멀티프로세스 방식은 언제 사용하나요?

**Reply by SionBackEnd** [reply]
@kkg5  저도 궁금합니다.

**Reply by kkg5** [reply]
간단하게 작업관리자에 각각 다른 PID로 크롬이 여러개 있는 걸 보면 멀티 프로세싱을 사용하긴 하는 것 같네요.
좀 더 찾아봐야겠습니다ㅎㅎ

![image](https://github.com/Team-Psyduck/tech-sharing/assets/97646802/9210d0ae-fea7-45f7-88e2-ccfd68768d89)

**Reply by kkg5** [reply]
```스레드는 하나의 프로세스 안에서 작동하기 때문에 하나의 스레드가 죽으면 전체 프로세스에 영향을 끼치게 된다.``` [출처](https://iambeginnerdeveloper.tistory.com/186)

크롬을 사용하다가 하나의 탭이 멈춘 경우에 다른 탭을 사용할 수 있었던 경험이 있는데
멀티 프로세스 방식이라 가능했던 것 같네요.

멀티 스레드와 달리 서로 영향을 주지 않을 수 있다는 장점이 있는 것 같습니다ㅎㅎ

**Comment by eatingbug** [comment]
멀티프로세스는 프로세스 간의 컨텍스트 스위칭 시 캐시메모리에 대한 데이터까지 초기화되므로 오버헤드가 크다. 
하지만 멀티스레드는 프로세스 내의 메모리를 공유하기 때문에 스레드간 데이터를 주고 받는것이 간단해지고 시스템 자원 소모가 줄어들게 된다.

다만 멀티스레드 방식은 스레드 간의 자원공유를 위해 전역변수를 사용하므로 동기화 문제나 Deadlock 과 같은 문제가 발생할 수 있다.

**Comment by SionBackEnd** [comment]
멀티 프로세스 단일 스레드 방식은 단일 프로세스 멀티 스레드 방식보다 컨텍스트 스위치 비용이 더 들게 됩니다.
